Record `tsc` compiled code

dustinbyrne opened the pull request:
This fixes an issue found while trying to record @appland/scanner and adds some additional test cases.

Resolves #173

lachrist commented:
Hi @dustinbyrne thanks for digging into this. The trace post-processing is rather complex and I remember that the pipeline of modules relies on assertion made earlier. So I don't think you can simply silence assertions. However I agree that producing something is better than panicking. I think the best way to proceed is to try the complex event re-ordering. If it fails (via assertions) we can try a simpler post-processing where we simply make sure the produced appmap is valid but we do not make any guarantee on the order of events.

dustinbyrne commented:
There are two items blocking this:

First, all AppMaps recorded from Jest tests are empty. This may be related to #172 where a transform is being applied. After all, this is using `ts-jest`. It may also be a configuration issue, but `process` recordings are functional. TBD.

Lastly, Jest command line arguments are not being transformed safely, and I've yet to come up with a strategy which doesn't involve mirroring Jests command line logic. Here are some examples where it breaks down.
```sh
$ npx jest authn
```
Would result in the following transformed command:
```sh
$ npx jest --runInBand --setupFilesAfterEnv /module/path.js authn
```
This is incorrect because `authn` will be read in as a setup file instead of a test path pattern. To retain the same behavior, the command would be one of the following:
```sh
$ npx jest --runInBand --setupFilesAfterEnv /module/path.js -- authn
$ npx jest --runInBand --setupFilesAfterEnv /module/path.js --testPathPattern authn
```
This is complicated by the fact that the user may add other options to the command line:
```
$ npx jest --detectOpenHandles authn
```
Would need to resolve to one of the following:
```
$ npx jest --runInBand --setupFilesAfterEnv /module/path.js --detectOpenHandles -- authn
$ npx jest --runInBand --setupFilesAfterEnv /module/path.js --detectOpenHandles --testPathPattern authn
```
This is again complicated by the fact that we don't know if `authn` is a test path pattern or an argument of `detectOpenHandles`. Any tokens remaining after parsing options are treated as test path patterns. Ideally, we would not want to re-implement this logic as it's likely to change (and probably already has across versions).

lachrist commented:
Ah I did not expect `--setupFilesAfterEnv` to eats all the non-dashed argument after it. The fix with the transformer will require to parse both the configuration file and the arguments. I will make sure I take this into account.

dustinbyrne commented:
The transformer fix is in the works and I've opened a separate issue for the Jest command line arguments (https://github.com/getappmap/appmap-agent-js/issues/179).

I don't think it's worth waiting for the outstanding items before shipping these.

lachrist reviewed (changes requested):
Thanks for the work. Some information on the failures you had related the two last commits would help. Cheers!

dustinbyrne commented on components/agent/default/source-map.mjs:
```diff
@@ -23,6 +25,19 @@ export const loadSourceMap = (file) => {
       );
       return createMirrorSourceMap(file);
     }
+    const match = maybe_url.match(/data:(.*);(.*),(.*)/u);
+    if (match) {
+      const [, contentType, encoding, data] = match;
+      if (contentType === "application/json") {
+        content = Buffer.from(data, encoding).toString("utf-8");
+      } else {
+        logWarning(
+          "Unsupported source-map content type in %j: %j",
+          maybe_url,
+          contentType,
+        );
+      }
+    }
```
It wasn't clear to me that `readFile` was also parsing and returning the content of `data` URLs. I was probably experiencing some other sort of misconfiguration at the time. In any case, I've pulled the change out but left the tests in.

dustinbyrne commented on components/trace/appmap/classmap/entity.mjs:
```diff
@@ -91,24 +92,29 @@ export const getEntityLabelArray = (entity) =>
   if (entity.type === "class") {
     return entity.name;
-  } else if (entity.type === "function") {
-    assert(
-      maybe_parent_entity !== null,
-      "function entity at root level",
-      InternalAppmapError,
-    );
```
If `wrapRootEntityArray` is responsible for ensuring functions have a parent, I'd assume this assertion is also made there. Enforcing an assertion solely downstream causes the root issue to become opaque and difficult to debug once the assertion is broken. 

I'm not really sure _why_ this is happening, but this assertion is what's causing the agent to panic. The resulting AppMap is valid, so from an end-user experience, what's best? I'd argue that a clean exit code and the getting the resulting data out is a much better experience than a panic. Most users won't take the time to report the issue, they'll just move on.

lachrist replied:
If you look at the content of `wrapRootEntityArray`, you will see that this assertion will always be verified. That that there is is impossible the invoke this function and break this assertion. I only use assertions when the blame can be put on the consumer of the method. Otherwise you pretty much fill up your code with assertions. But this is some general philosophical mombo-jumbo.

Of course it is better to produce a valid appmap than crashing. I usually only add assertions when subsequent operations do not make sense if the assertion is failing. In this case, I added the assertion because the parent code entity was needed to compute the qualified name of the function.

Anyhow, I found the root cause of the issue: wrapping root-level functions was done *before* exclusions. https://github.com/getappmap/appmap-agent-js/pull/185

dustinbyrne commented on components/hook-group/node/index.mjs:
```diff
@@ -12,7 +10,14 @@ export const hook = (agent, { ordering }) => {
+        // However, the asyncId is guaranteed to be unique so we should be able to safely ignore it if it's
+        // already been seen.
+        /* c8 ignore next 3 */
+        if (groups.has(id)) {
+          return;
+        }
```
I guess it's up to you. I've never really heard of guard clauses being considered as an anti-pattern.

lachrist replied:
I'm not saying it is bad or anything. I'm just saying that you won't find this pattern anywhere else in the codebase. This is just about consistency. This is not a hill that I'm willing to die for, so please do whatever you feel is best.

lachrist reviewed (changes requested):
I still have an issue with your removal of the assertion. I made a fix in https://github.com/getappmap/appmap-agent-js/pull/185. You could leave the test but it will need to be updated because the qualified name of a root function is with the filename.

dustinbyrne commented:
I've rebased off the branch in #185 and dropped the commit to pass the assertion. The test seems unnecessary without providing an undefined or null parent.

lachrist reviewed (approved):
Brazinga

appland-release commented:
:tada: This PR is included in version 13.1.2 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@appland/appmap-agent-js/v/13.1.2)
- [GitHub release](https://github.com/getappmap/appmap-agent-js/releases/tag/v13.1.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
